### PR TITLE
Provide 2 types of scoped_session in DI (initially "type of session parameter should be scoped_session")

### PR DESCRIPTION
**Pull request by jdavcs**
**UPDATE:** Initially, this was fixing one bug, as a step towards migrating to SQLAlchemy 2.0. However, [as described in this comment](https://github.com/galaxyproject/galaxy/pull/12931#issuecomment-971726107), the bug fix exposed a deeper issue. Since one cannot be fixed without the other, this PR contains both fixes.

**Summary of problem being solved:**
The dependency injection system stores only one scoped_session. We need 2: one for the main model, the other one for the tool_shed_install model. ([see comment below for details](https://github.com/galaxyproject/galaxy/pull/12931#issuecomment-971726107))

**Implementation summary:**
I've added 2 empty classes that extend `sqlalchemy.orm.scoped_session`. I use these types as keys in the DI to store `galaxy.model.context` and `galaxy.install_model.context` - which are scoped_session objects, each bound to the appropriate engine. Thus, we simply specify one or the other in a constructor - and the right scoped_session will be injected.
(at first, I used these in model.base to instantiate one or the other - but then realized this was completely unnecessary!)

I tried placing these class definitions in `model.base.py`, as well as `model.__init__.py` (which caused a circular dependency and raised an error) and `model.mapping.py`. In the end I created a separate module, and this felt like the cleanest/most intuitive location. But this is completely subjective. 

Naming: I wanted something succinct, and using `gxy_` and `tsi_` prefixes (for galaxy and tool_shed_install) felt right (all the alternatives I came up with seemed long and awkward). I'm also using these all over the new migrations code (#12869), so these will be consistent. But again, very subjective.

I haven't looked too deeply into Lagom - so maybe there is a more elegant way to address such a situation (the need for storing 2 *different* objects of the *same* type - I think we considered the possibility of such a case when we first discussed adding DI).

---------------------
**INITIAL DESCRIPTION:**
(this is a step towards #12541)

The type of the `session` parameter should be scoped_session, not sessonmaker. With the former, the injector locates the instance of scoped_session that has been [bound to an engine here](https://github.com/galaxyproject/galaxy/blob/dev/lib/galaxy/model/base.py#L55) and [registered here](https://github.com/galaxyproject/galaxy/blob/dev/lib/galaxy/app.py#L135). With the latter, DI cannot find an instance of `sessionmaker`, so it creates a new instance (that uses default parameters).

**Why this is a problem (besides being incorrect)**
The session we get on the next line is a session object that has been constructed using default arguments. Most importantly, it does not have a bind that we set in `model/base`. This error went unnoticed because we also [bind the model's `MetaData` property to the same engine](https://github.com/galaxyproject/galaxy/blob/dev/lib/galaxy/model/tool_shed_install/mapping.py#L15). Given this bind, having no bind on the session object is inconsequential: the query will still execute. 
However, this will change as soon as we remove the bind from the metadata, which we have to do because bound metadata is [no longer supported in SQLAlchemy 2.0](https://docs.sqlalchemy.org/en/14/changelog/migration_20.html#implicit-and-connectionless-execution-bound-metadata-removed).

```
# this is the correct/fixed version (note the session properties):
(Pdb) session
<sqlalchemy.orm.scoping.scoped_session object at 0x7f01d00cff40>
(Pdb) pp session.__dict__
{'current': <sqlalchemy.orm.scoping.scoped_session object at 0x7f01d00cff40>,
 'registry': <sqlalchemy.util._collections.ScopedRegistry object at 0x7f01d00cf9a0>,
 'session_factory': sessionmaker(class_='Session', 
bind=Engine(sqlite:////home/sergey/0dev/galaxy/_galaxy/dev/database/universe.sqlite?isolation_level=IMMEDIATE), autoflush=False, autocommit=True, expire_on_commit=True

# this is the incorrect version (note the session properties)
(Pdb) session
sessionmaker(class_='Session', bind=None, autoflush=True, autocommit=False, expire_on_commit=True)
(Pdb) pp session.__dict__
{'class_': <class 'sqlalchemy.orm.session.Session'>,
 'kw': {'autocommit': False,
        'autoflush': True,
        'bind': None,
        'expire_on_commit': True}}
```

## How to test the changes?
(Select all options that apply)
- [x] This is a refactoring of components with existing test coverage.

## License
- [x] I agree to license these contributions under [Galaxy's current license](https://github.com/galaxyproject/galaxy/blob/dev/LICENSE.txt).
- [x] I agree to allow the Galaxy committers to license these and all my past contributions to the core galaxy codebase under the [MIT license](https://opensource.org/licenses/MIT). If this condition is an issue, uncheck and just let us know why with an e-mail to galaxy-committers@lists.galaxyproject.org.

### CODE REVIEW

**Review by mvdbeek — APPROVED** [review]
Thanks!

**Comment by mvdbeek** [comment]
Hmm, the toolshed tests timed out, that might be a real issue.

**Comment by ic4f** [comment]
> Hmm, the toolshed tests timed out, that might be a real issue.

Looking into it.

**Comment by ic4f** [comment]
> > Hmm, the toolshed tests timed out, that might be a real issue.
> 
> Looking into it.

@mvdbeek Thanks - it was directly related. It's an issue I was going to address after this, but this won't work without it, so I'll address it here. Here's what's happening: in `ToolShedRepositoryCache.rebuild()` when SQLAlchemy tries to [execute this query](https://github.com/galaxyproject/galaxy/blob/dev/lib/galaxy/tools/cache.py#L303), it looks for a connectable (stored in a bind). Previously, the bind on the session was None (due to the bug addressed here), so it looked in `ToolShedRepository.metadata` - and found the correct engine. However, now that the session *has* a bind, the engine from that bind is what gets used. But that engine may be for a database that does not contain the install_model tables! Which is what happens in the test (because `install_database_connection` is set to a different db). 

In summary, the root cause and the issue that needs to be addressed is this: we store **one** session with **one** bind in DI, but **we need two** - which comes into play when `install_database_connection` is set to a different database. The issue went unnoticed because the distinction between the 2 databases was masked by the combination of bound metadata (which won't work in SA 2.0) and the sessionmaker bug (which is being fixed here).

I'm working on it.

**Comment by ic4f** [comment]
I've updated the PR description detailing the new changes.

File: lib/galaxy/model/scoped_session.py [hunk 1]
```diff
@@ -0,0 +1,15 @@
+"""
+These classes are used for registering different scoped_session objects with
+the DI framework. This type distinction is necessary because we need to store
+scoped_session objects that produce sessions that may have different binds
+(i.e., if the tool_shed_install model uses a different database).
+"""
+from sqlalchemy.orm import scoped_session
+
+
+class gxy_scoped_session(scoped_session):
```
Comment (mvdbeek) [inline lib/galaxy/model/scoped_session.py]: Can you rename this to `galaxy_scoped_session` or just `scoped_session` and `tsi_scoped_session` to `install_model_scoped_session` ?

**Review by mvdbeek — APPROVED** [review]

File: lib/galaxy/model/scoped_session.py (hunk 1 above)
Comment (jdavcs) [reply lib/galaxy/model/scoped_session.py]: Done! `galaxy_scoped_session` and `install_model_scoped_session`.

**Comment by jmchilton** [comment]
Hmm... I'm okay with this solution. I would prefer if we just put one scoped session in the DI container and then put an explicit ToolCache in the container with the other session manually created and pass along. If we were explicit about that we wouldn't need lagom to wire this up for us and get it wrong. But... maybe that is harder to do than I'm aware and there may be times down the road being more explicit here clears things up.

**Comment by ic4f** [comment]
> Hmm... I'm okay with this solution. I would prefer if we just put one scoped session in the DI container and then put an explicit ToolCache in the container with the other session manually created and pass along. If we were explicit about that we wouldn't need lagom to wire this up for us and get it wrong. But... maybe that is harder to do than I'm aware and there may be times down the road being more explicit here clears things up.

Thanks for looking into this! I fear we would have to manually create a session for any of the 6 models in the install_model in that case (whenever we need to touch the install_model database), since their metadata property will no longer be bound to the engine. 

That said, a session *may* have multiple binds, and *it is* possible to make the session factory (`sessionmaker`, in our case) in the scoped_session produce session objects with the correct bind (It is not straightforward: unlike with the session constructor, you can only pass a single bind to the session factory, but we need two.. One way I found was to use the `info` argument, which is a dictionary of arbitrary user data - so ..not straightforward). But that would mean we need the same session on both `ModelMapping` objects: for the galaxy model and the install model - which, I think, would lead to changing how we create and handle scoped_session objects.

So, thanks for being OK with the current version :smiley: ! I'm happy to revisit this implementation right after we launch alembic, which will introduce its own changes. We can then reexamine this.

**Comment by meadsteve** [comment]
Author of Lagom here. Regarding this point:

> I haven't looked too deeply into Lagom - so maybe there is a more elegant way to address such a situation (the need for storing 2 different objects of the same type - I think we considered the possibility of such a case when we first discussed adding DI).

For your use case defining new types is actually what I use myself. When designing lagom I wanted things to be represented by types rather than magic strings or decorators so that IDEs and other tooling could understand what's happening. 

The alternative if a new type doesn't make sense is to provide lagom with lambda construction functions for the dependencies that need differently configured instances of a dependency. Like this:

```python
container[SomeRepoOne] = lambda: SomeRepoOne(db = Database.data_centre_1)
container[SomeRepoTwo] = lambda: SomeRepoTwo(db = Database.data_centre_2)

container[Database] = UnresolvableTypeDefinition("DI can not automatically resolve the DB. You must provide explicit construction")
```

However I would always lean towards doing the type based one if possible as it communicates more information in my opinion.

**Comment by ic4f** [comment]
@meadsteve Thank you for looking into this and confirming that our approach is reasonable! The alternative you suggest is definitely helpful too - it's very possible we'll encounter that kind of use case, so we'll lean on that option if need be.
